File: bot/web/lib/src/tic_tac_toe/players.dart
import "../users.dart";
import "../utils.dart";

class Player {
  final User user;

  Player(Map<String, dynamic> data) : user = User(data["user"]);

  static Pair<Player, Player?> pairFromJson(List<Map<String, dynamic>?> data) => Pair<Player, Player?>(
        Player(data[0]!),
        data[1] == null ? null : Player(data[1]!),
      );
}

File: bot/web/scripts/tic-tac-toe/players.ts
/// <reference path="../discord/users.ts" />


namespace tic_tac_toe {
    export class Player {
        public readonly user: discord.User | null;

        private constructor(user: discord.User | null) {
            this.user = user;
        }

        public get displayName(): string {
            if (this.user !== null) return this.user.name;
            return "Guest";
        }

        public static fromObject(data: object): Player {
            return new Player(data["user"] !== null ? discord.User.fromObject(data["user"]) : null);
        }
    }
}